numpy/_core/src/umath/string_fastsearch.h

```diff
@@ -62,7 +62,7 @@ findchar(Buffer<enc> s, Py_ssize_t n, npy_ucs4 ch)
     Buffer<enc> p = s, e = s + n;
 
     if (n > MEMCHR_CUT_OFF) {
-        p = s.buffer_memchr(ch, n);
+        p = p.buffer_memchr(ch, n);
```

Comment:
This solution works, but it sort-of works around the fact that `buffer_memchr` updates the position of the buffer in-place. My C++ skills were insufficient to trivially solve that.

Reply:
That's true. I think it'd be preferrable to fix `buffer_memchr` to return a new buffer without side effects on the calling one.

This would do it probably:

```c
inline Buffer<enc>
buffer_memchr(npy_ucs4 ch, int len)
{
    Buffer<enc> newbuf = *this;
    switch (enc) {
    case ENCODING::ASCII:
        newbuf.buf = (char *) memchr(buf, ch, len);
        break;
    case ENCODING::UTF32:
        newbuf.buf = (char *) wmemchr((wchar_t *) buf, ch, len);
        break;
    }
    return newbuf;
}
```

Reply:
Yes, that makes more sense, and seems to work.

Reply:
OK, the first commit now changes `buffer_memchr` like you suggested.